R/join_keys.R

```
@@ -0,0 +1,668 @@
+    dataset_1 <- names(join_keys_obj)[dataset_1]
+  }
+
+  # When retrieving a relationship pair, it will also return the symmetric key
+  new_jk <- new_join_keys()
+  queue <- dataset_1
+  bin <- character(0)
+
+  # Need to iterate on a mutating queue if subset of a dataset will also
+  #  select its parent as that parent might have relationships with others
+  #  already selected.
+  while (length(queue) > 0) {
+    ix <- queue[1]
+    queue <- queue[-1]
+
+    if (ix %in% bin) {
+      next
+    }
+    bin <- c(bin, ix)
+
+    ix_parent <- parent(join_keys_obj, ix)
+
+    if (checkmate::test_string(ix_parent, min.chars = 1) && !ix_parent %in% c(queue, bin)) {
+      queue <- c(queue, ix_parent)
+    }
+
+    ix_valid_names <- names(join_keys_obj[[ix]]) %in% c(queue, bin)
+    if (keep_all_foreign_keys) {
+      ix_valid_names <- rep(TRUE, length(names(join_keys_obj[[ix]])))
+    }
+
+    new_jk[[ix]] <- join_keys_obj[[ix]][ix_valid_names]
+
+    # Add primary key of parent
+    if (length(ix_parent) > 0) {
+      new_jk[[ix_parent]][[ix_parent]] <- join_keys_obj[[ix_parent]][[ix_parent]]
+    }
+  }
+
+  common_parents_ix <- names(parents(join_keys_obj)) %in% names(new_jk) &
+    parents(join_keys_obj) %in% names(new_jk)
+
+  if (any(common_parents_ix)) parents(new_jk) <- parents(join_keys_obj)[common_parents_ix]
+
+  new_jk
+}
+
+#' @rdname join_keys
+#' @details
+#' Setter via index directly (bypassing the need to use `join_key()`).
+#' When `dataset_2` is omitted, it will create a primary key with `dataset_2 = dataset_1`.
+#'
+#' @param value (`character` vector) value to assign.
+#'
+#' @export
+#'
+#' @examples
+#'
+#' # Setter via index ----
+#'
+#' jk <- join_keys(
+#'   join_key("ds1", "ds2", "col12"),
+#'   join_key("ds3", "ds4", "col34")
+#' )
+#'
+#' # overwrites previously defined key
+#' jk["ds1"] <- list(ds2 = "(new)co12")
+#' jk["ds1"] <- list(ds3 = "col13", ds4 = "col14")
+#' jk
+#'
+#' jk[c("ds1", "ds2")] <- list(ds5 = "col*5")
```

this is bad in my opinion. We shouldn't support this

```r
# a <- list(a = 1)
# a[c("a", "b")] <- list(IGNORED_NAME = "hello")
# a
# $a
# [1] "hello"
# $b
# [1] "hello"
```

---

Should we support any (single bracket) `[<-` operation? They all seem problematic.

*And they are not overwriting, which they should be.

---

In other words, this should throw an error and point to the right tools `[[<-`

---

I don't think we should. If `base` doesn't have a clear answer we can do following. 

```r
[<-.join_keys` <- function(x, i, value) {
  stop("Can't use `[<-` for object `join_keys`") 
}
```

As I understand, above shouldn't actually break anything.

---

Or remove method at all - [<-.list would take over. We don't need to support this on. our side.

---

`[<-.list` will wreck havoc on  join_keys structure. I'm tending towards throwing error.

---

Done in acc5454a